### DISCUSSION TITLE
Textfield styling

### DISCUSSION THREAD

**Discussion by beppo-ivel** [appcelerator/titanium_mobile#13080] [Q&A] [answered]
I have several questions with respect to styling [text field](https://docs.appcelerator.com/platform/latest/#!/api/Titanium.UI.TextField) for android:

1) How to change the cursor color?
2) How to hide the underscore of android text fields or set the underscore in the same color as background to make it invisible? 
3) How to set the top- and bottom margin between elements. For example if you add an element then you have to re-define the the top and bottom values of the elements. Maybe that could be avoid by using margin properties which organizes the space between elements?
4) Is it possible to create a border with small description text as you can see [here](https://stackoverflow.com/questions/45150781/placing-label-text-inside-the-border-of-a-text-input)?

**Comment by m1ga** [comment, marked as answer]
1: guess you have to use a Android theme for that. Never done that but if you search for it you'll find some examples. This might be a starting point: https://www.11zon.com/zon/android/how-to-change-android-edittext-cursor-color.php (its not a theme but the property name might be the same)
2: https://titaniumsdk.com/api/titanium/ui/textfield.html#borderstyle
3: use a class with the top/bottom properties? There was a discussion about some kind of `margin`but it is not implemented.
4: https://titaniumsdk.com/api/titanium/ui/textfield.html#hinttype

I recommend checking out kitchensink from https://github.com/appcelerator/kitchensink-v2/ especially the screenshots in https://github.com/appcelerator/kitchensink-v2/pull/56

**Reply by jquick-axway** [reply]
@m1ga is right. I recommend that you update to Titanium 10.0.0 or higher and set the `TextField`/`TextArea` "borderStyle" [property to [Ti.UI.INPUT_BORDERSTYLE_ROUNDED](https://titaniumsdk.com/api/titanium/ui.html#input_borderstyle_rounded). This will exactly match the look and feel you see in Google's own apps such as "Contacts". 

It's best to control the colors via an XML theme as shown here. Note that the "values" folder is the default light theme and "value-night" is the dark theme. I recommend keeping it simple as shown in the below sample project.
https://github.com/appcelerator/kitchensink-v2/blob/master/app/platform/android/res/values/custom_theme.xml
https://github.com/appcelerator/kitchensink-v2/blob/master/app/platform/android/res/values-night/custom_theme.xml

Google documents which theme color constants are used per TextField component below (search for "Container Attributes" section). Notice they use different shades of "colorPrimary" and "colorOnPrimary" styles.
https://material.io/components/text-fields/android#outlined-text-field

**Reply by beppo-ivel** [reply]
> 1: guess you have to use a Android theme for that...

I'm confused about how to implement themes. I read out both articles [here](https://docs.appcelerator.com/platform/latest/#!/guide/Alloy_Styles_and_Themes-section-src-35621526_AlloyStylesandThemes-Themes) and [here](https://docs.appcelerator.com/platform/latest/#!/guide/Android_Themes). I think the general concept is clear, that themes provide a way to overwrite or modify files for a specific brand of your app. But in detail I have two question:
1) What kind of style can I change in my theme what I can not do without?
2)  Why is [platform/android/res/values/colors.xml](https://docs.appcelerator.com/platform/latest/#!/guide/Android_Themes) not located under themes?

> 2: https://titaniumsdk.com/api/titanium/ui/textfield.html#borderstyle 
The property borderColor (green border) has no effect to that border (blue border), which is styled by the property borderStyle. 
![txtfielld](https://user-images.githubusercontent.com/72635809/136253808-08756b13-6372-431e-bfa5-b9a8c0cafbc2.png)


@jquick-axway: I updated the components:
```
Node.js
  Node.js Version             = 14.17.6
  npm Version                 = 6.14.15

Titanium CLI
  CLI Version                 = 5.3.0

Titanium SDK
  SDK Version                 = 10.0.0.GA
```

**Reply by m1ga** [reply]
> 
    What kind of style can I change in my theme what I can not do without?
    Why is platform/android/res/values/colors.xml not located under themes?

That's very Android/native specific. Structure: it's just like this in the native world :) I don't know why, never questioned it :rofl: 
You can style almost everything with a theme. Again: native world. Like spacing between letters, the Actionbar, the color of the statusbar and so on.

**Reply by m1ga** [reply]
> 2: https://titaniumsdk.com/api/titanium/ui/textfield.html#borderstyle
> The property borderColor (green border) has no effect to that border (blue border), which is styled by the property borderStyle.

I guess you've set the green border in Titanum, right? If you want to style using a theme have a look at the material.io link posted earlier.

You'll end up with something like this
```
	<style name="defaultTheme" parent="Theme.MaterialComponents.DayNight">
		<item name="textInputStyle">@style/Widget.App.TextInputLayout</item>
	</style>
	
	<style name="Widget.App.TextInputLayout" parent="Widget.MaterialComponents.TextInputLayout.OutlinedBox">
		<item name="hintTextColor">#00ff00</item>
		<item name="boxStrokeColor">#ff0000</item>
	 <item name="boxStrokeWidth">2dp</item>
	</style>
```
![Screenshot_20211006-201125](https://user-images.githubusercontent.com/4334997/136259718-a1b20dd0-7a51-4dd7-ae65-a287230b1751.png)

**Reply by beppo-ivel** [reply]
> Again: native world.

What I don't get is why is there a abstraction layer in titanium to style components via *.tss files and themes to define styles for several brands of the app. And why are there some components like border color of a textfield where I have to use a android native way to control their style?

**Reply by m1ga** [reply]
since that is very specific to Android you'll have to do it using a theme (like in a native app) or extend the SDK and make that property available like border for a normal view. 

That "move text up" feature was introduced very recently. You can build that yourself with labels, views and a plain textfield. Then you can use your normal TSS properties to style everything. But if you want to use the Material Textfield you style it the Material way :-)

**Reply by jquick-axway** [reply]
> What I don't get is why there is a abstraction layer in titanium to style components via *.tss files...

Because Google doesn't offer Java APIs to change every aspect of their UI components. Some things they do, like background colors or text colors, but not everything. For colors that can't be changed via Google's Java APIs, your only option is to do so via theme XML files. Bottom line, this is how Google wants you to do it. You're really complaining about Google's UI framework, not Titanium.

But the good news is handling colors via a theme XML file makes it easy to support dark/light mode on Android. It's easier than iOS in my opinion. What I like to do is only set colors in the .tss file for iOS, but never set the colors on Android and instead leverage the the theme XML files. The kitchensink-v2 app is a good example of this.

**Side Note:** Android 12's upcoming "Material You" feature which has the end-user choose the color theme that is to be used by all apps (such as red, green, blue, etc.) is another good reason why this should really be handled by native theme XML files as well.

**Reply by beppo-ivel** [reply]
> You can build that yourself with labels, views ....

Does the same applies for title of a tab group (actionbar?)? Because I'm not able to change the color of tab group title via *.tss. And I don't understand why I can define a title for each window but when I want to have different titles in each tab I have to set a new tab group title via (or is there also a other possibility?):

```
$.tabGroup.addEventListener('focus', function(e) {
	        let title = $.tabGroup.activeTab.window.title;
        	$.tabGroup.activity.actionBar.title = title;
	});
```
@jquick-axway 
> You're really complaining about Google's UI framework, not Titanium.

Was not my intention to complaining the titanium framework it was only gap of my understanding.

**Reply by m1ga** [reply]
styling: yes check e.g. https://stackoverflow.com/a/5881739/5193915 (first google hit, didn't want to search longer or in my files :smile: )

**Reply by m1ga** [reply]
not sure if I got your title question right but:
```js
var window1 = Titanium.UI.createWindow({
	title: 'Tab 1'
});

var window2 = Titanium.UI.createWindow({
	title: 'Tab 2'
});

var window3 = Titanium.UI.createWindow({
	title: 'Tab 3'
});

var tab1 = Ti.UI.createTab({
	title: 'Tab 1',
	window: window1
});

var tab2 = Ti.UI.createTab({
	title: 'Tab 2',
	window: window2
});

var tab3 = Ti.UI.createTab({
	title: 'Tab 3',
	window: window3
});

tab1.addEventListener('selected', function() {
	tabGroup.title = 'Tab 1';
});

tab2.addEventListener('selected', function() {
	tabGroup.title = 'Tab 2';
});

tab3.addEventListener('selected', function() {
	tabGroup.title = 'Tab 3';
});

var tabGroup = Titanium.UI.createTabGroup({
	tabs: [tab1, tab2, tab3],
	swipeable: true,
	exitOnClose: false,
	style: Titanium.UI.Android.TABS_STYLE_BOTTOM_NAVIGATION,
	shiftMode: false
});

tabGroup.open();
```
changes the tabgroups title

**Reply by beppo-ivel** [reply]
> first google hit,...

ok ok I know but I was confused because I was able to change the color of the actionbar via the window style property `barColor` but was not able to set the color of the object in *.tss or neither via `$.tabGroup.activity.actionBar.color/titleColor`. So it's confusing in the beginning where to find the right control.

**Reply by m1ga** [reply]
yes, can be confusing if you start with app development. Just remember: when it's not in the documentation as a property (and Android only) you might want to search for `change component-xyz in theme android`. At least that's how I do it :rofl:

**Comment by jquick-axway** [comment]
All the subfolders under your Titanium project's `./platform/android/res` folder does what Google documents here...
https://developer.android.com/guide/topics/resources/providing-resources

Follow the kitchensink-v2 example links I posted above. In that example, the app uses a "red" theme, which includes the text field border theme. It's simple. Just try it.

There's also a guide on how to set up themes below.
https://titaniumsdk.com/guide/Titanium_SDK/Titanium_SDK_How-tos/User_Interface_Deep_Dives/Android_UI_Components_and_Conventions/Android_Themes.html
